### PR TITLE
Clean up block/unblock handling of resources in new scheduler

### LINKED ISSUE #12937: [Core][New Scheduler] Actors borrowing CPUs infinitely.
### What is the problem?

```python3
    ray.init(
        num_cpus=1,
        _system_config={
            "debug_dump_period_milliseconds": 500
        })
    @ray.remote(num_cpus=1)
    class Foo:
        def f(self):
            return 0
    @ray.remote
    def f():
        # Creating both actors is not possible.
        actors = [Foo.remote() for _ in range(3)]
        for a in actors:
            ray.get(a.f.remote())
    # Run in a task to check we handle the blocked task case correctly
    f.remote()
```
This job is supposed to be hanging because worker can borrow cpu to an actor once (and the subsequent actor creation cannot borrow a cpu anymore), but it seems like it succeeds when I ran this script.

It fails (which is the expected behavior) when I run it with the old scheduler.

But when we run it with only tasks, it works for both cases (which is the expected behavior)
```python3
    ray.init(
        num_cpus=1,
        _system_config={
            "debug_dump_period_milliseconds": 100000
        })

    @ray.remote(num_cpus=1)
    def g():
        time.sleep(2)
        return 0

    @ray.remote
    def f():
        for _ in range(2):
            ray.get(g.remote())

    ray.get(f.remote(), timeout=5)
```

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This simplifies the block/unblock problem by allow available_resources to go negative in the new scheduler (overcommit). Previously this wasn't allowed, which made it difficult to avoid two edge conditions: (1) infinite cpu acquisition by repeatedly block/unblocking a task, and (2) available_resources sometimes exceeding total_resources, requiring tracking of the overflow.

By allowing available_resources to go negative, we can maintain the invariant that `available_resources <= total_resources`, and trivially prevent infinite resource acquisition.

Closes https://github.com/ray-project/ray/issues/12937

### CODE REVIEW

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -956,8 +956,8 @@ const std::vector<std::shared_ptr<WorkerInterface>> WorkerPool::GetAllRegistered
 }
 
 void WorkerPool::WarnAboutSize() {
-  for (const auto &entry : states_by_lang_) {
-    auto state = entry.second;
+  for (auto &entry : states_by_lang_) {
+    auto &state = entry.second;
```
Comment (ericl) [inline src/ray/raylet/worker_pool.cc:960]: Bug: implicit copy of mutable state :(

File: src/ray/raylet/node_manager.cc [hunk 2]
```diff
@@ -2189,9 +2191,11 @@ void NodeManager::HandleDirectCallTaskUnblocked(
       cpu_instances = worker->GetAllocatedInstances()->GetCPUInstancesDouble();
     }
     if (cpu_instances.size() > 0) {
-      new_resource_scheduler_->SubtractCPUResourceInstances(cpu_instances);
-      new_resource_scheduler_->AddCPUResourceInstances(worker->GetBorrowedCPUInstances());
-      worker->ClearBorrowedCPUInstances();
+      // Important: we allow going negative here, since otherwise you can use infinite
+      // CPU resources by repeatedly blocking / unblocking a task. By allowing it to go
+      // negative, at most one task can "borrow" this worker's resources.
+      new_resource_scheduler_->SubtractCPUResourceInstances(
```
Comment (rkooo567) [inline src/ray/raylet/node_manager.cc:2197]: Just curious, but what's happening if the RPC failed?

File: src/ray/raylet/scheduling/cluster_task_manager.cc [hunk 3]
```diff
@@ -242,10 +243,7 @@ void ClusterTaskManager::TasksUnblocked(const std::vector<TaskID> ready_ids) {
       const auto &scheduling_key = task.GetTaskSpecification().GetSchedulingClass();
       RAY_LOG(DEBUG) << "Args ready, task can be dispatched "
                      << task.GetTaskSpecification().TaskId();
-      // Note: we transition tasks back to the scheduling queue instead of directly
-      // to dispatch. This allows AnyPendingTasks() to simply check the scheduling
-      // queue to see if any tasks are blocked on resource availability: see #12438
-      tasks_to_schedule_[scheduling_key].push_back(work);
+      tasks_to_dispatch_[scheduling_key].push_back(work);
```
Comment (ericl) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:246]: Revert unnecessary logic from last pr

**Review by wuisawesome — APPROVED** [review]

**Comment by ericl** [comment]
Java tests failing with same block/unblock issue.

**Pull request merged**